tests/phpunit/payment-methods/test-wc-stripe-express-checkout-helper.php

```diff
@@ -205,6 +205,61 @@ public function test_hides_ece_if_stripe_gateway_unavailable() {
 		WC()->payment_gateways()->payment_gateways = $original_gateways;
 	}
 
+	/**
+	 * Test should_show_express_checkout_button, filter logic.
+	 */
+	public function test_shows_toggle_ece_when_forcing_the_filter_value() {
+		$this->set_up_shipping_methods();
+
+		$wc_stripe_ece_helper_mock = $this->createPartialMock(
+			WC_Stripe_Express_Checkout_Helper::class,
+			[
+				'is_product',
+				'allowed_items_in_cart',
+				'should_show_ece_on_cart_page',
+				'should_show_ece_on_checkout_page',
+			]
+		);
+		$wc_stripe_ece_helper_mock->expects( $this->any() )->method( 'is_product' )->willReturn( false );
+		$wc_stripe_ece_helper_mock->expects( $this->any() )->method( 'allowed_items_in_cart' )->willReturn( true );
+		$wc_stripe_ece_helper_mock->expects( $this->any() )->method( 'should_show_ece_on_cart_page' )->willReturn( true );
+		$wc_stripe_ece_helper_mock->expects( $this->any() )->method( 'should_show_ece_on_checkout_page' )->willReturn( true );
+		$wc_stripe_ece_helper_mock->testmode = true;
+		if ( ! defined( 'WOOCOMMERCE_CHECKOUT' ) ) {
+			define( 'WOOCOMMERCE_CHECKOUT', true );
+		}
+		$original_gateways = WC()->payment_gateways()->payment_gateways;
+
+		// Add a non-taxable product to the cart.
+		$product = WC_Helper_Product::create_simple_product();
+		$product->set_virtual( false );
+		$product->set_tax_status( 'none' );
+		$product->save();
+
+		WC()->session->init();
+		WC()->cart->empty_cart();
+		WC()->cart->add_to_cart( $product->get_id(), 1 );
+
+		WC()->payment_gateways()->payment_gateways = [
+			'stripe'        => new WC_Gateway_Stripe(),
+			'stripe_alipay' => new WC_Gateway_Stripe_Alipay(),
+		];
+
+		// Should show by default if 'stripe' gateway is available, but hide anyway if the filter is set to false.
+		add_filter( 'wc_stripe_should_show_express_checkout_button', '__return_false' );
+		$this->assertFalse( $wc_stripe_ece_helper_mock->should_show_express_checkout_button() );
+
+		// Should hide by default if 'stripe' gateway is unavailable, but show anyway if the filter is set to true.
+		unset( WC()->payment_gateways()->payment_gateways['stripe'] );
+		add_filter( 'wc_stripe_should_show_express_checkout_button', '__return_true' );
+		$this->assertTrue( $wc_stripe_ece_helper_mock->should_show_express_checkout_button() );
+
+		// Restore original settings.
+		WC()->session->cleanup_sessions();
```

Comment:
Do we need/want to do `remove_filter(...)` as part of cleanup?

Reply:
Sure. Done in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4200/commits/43d09f09aff65d67f5ced24ae66d2bf441a39e0a